math.hpp

```diff
@@ -113,6 +133,262 @@ template<typename U, typename S, typename V, typename T>
+
+	static constexpr auto value   ()                  {return nominal() * fraction();}
+	static constexpr auto nominal ()                  {return tolerance_nom<quant_t        >{quant_t{tolerance_traits<float_t>::nominal ().v}};}
+	static constexpr auto fraction()                  {return tolerance_frc<fract_t,quant_t>{        tolerance_traits<float_t>::fraction().v };}
+};
+/** @} */
+
+/**
+ * @{
+ *
+ * @brief create a tolerance value, nominal, or fraction
+ *
+ * These function turn a value into a tolerance value, nominal, or fraction.
+ *
+ * @param val  value to turn into a tolerance value, nominal, or fraction
+ *
+ * @return tolerance value, nominal, or fraction
+ */
+template<typename T>             constexpr auto tolerance_value   (T val) {return tolerance_val<T  >{val};}
+template<typename T>             constexpr auto tolerance_nominal (T nom) {return tolerance_nom<T  >{nom};}
+template<typename T, typename F> constexpr auto tolerance_fraction(F frc) {return tolerance_frc<F,T>{frc};}
+/** @} */
+
+/**
+ * @{
+ *
+ * @brief get the default tolerance value, nominal, or fraction
+ *
+ * These function return the tolerance value, nominal, or fraction for a
+ * given type.
+ *
+ * @return default tolerance value, nominal, or fraction
+ */
+template<typename T>             constexpr auto tolerance_value   ()      {return tolerance_traits<T>::value   ();}
+template<typename T>             constexpr auto tolerance_nominal ()      {return tolerance_traits<T>::nominal ();}
+template<typename T>             constexpr auto tolerance_fraction()      {return tolerance_traits<T>::fraction();}
+/** @} */
+
+namespace detail {
+
+template<typename V, typename F, typename Q> constexpr auto get_tol_val(tolerance_nom<V>   n, tolerance_frc<F,Q> f) {return n * f;}
+template<typename V, typename F, typename Q> constexpr auto get_tol_val(tolerance_frc<F,Q> f, tolerance_nom<V>   n) {return get_tol_val(n,f);}
+template<typename V  >                       constexpr auto get_tol_val(tolerance_aux<tolerance_val_tag,V,V>     v) {return v;}
+template<typename V  >                       constexpr auto get_tol_val(tolerance_aux<tolerance_nom_tag,V,V>     n) {return get_tol_val(n, tolerance_traits<V>::fraction());}
+template<typename F, typename Q>             constexpr auto get_tol_val(tolerance_aux<tolerance_frc_tag,F,Q>     f) {return get_tol_val(f, tolerance_traits<Q>::nominal ());}
+
+template<typename T1, typename T2, typename Tol> constexpr bool is_near_impl   (T1 lval, T2 rval, tolerance_val<Tol> tol) {using std::abs; return abs(lval-rval) <= abs(tol.v);}
+template<typename T1, typename T2, typename Tol> constexpr bool is_smaller_impl(T1 lval, T2 rval, tolerance_val<Tol> tol) {using std::abs; return not is_near(lval,rval,tol) and lval-rval < abs(tol.v);}
+template<typename T1, typename T2, typename Tol> constexpr bool is_greater_impl(T1 lval, T2 rval, tolerance_val<Tol> tol) {using std::abs; return not is_near(lval,rval,tol) and lval-rval > abs(tol.v);}
```

Comment:
```suggestion
template<typename T1, typename T2, typename Tol> constexpr bool is_smaller_impl(T1 lval, T2 rval, tolerance_val<Tol> tol) {using std::abs; return lval-rval < -abs(tol.v);}
template<typename T1, typename T2, typename Tol> constexpr bool is_greater_impl(T1 lval, T2 rval, tolerance_val<Tol> tol) {using std::abs; return lval-rval > abs(tol.v);}
```

suffices.

Reply:
I think this would fail if negative values are passed in. However, there is no need to check for the tolerance twice. This is what I now have:

```suggestion
template<typename T1, typename T2, typename Tol> constexpr bool is_smaller_impl(T1 lval, T2 rval, tolerance_val<Tol> tol) {using std::abs; return not is_near_impl(lval,rval,tol) and lval < rval;}
template<typename T1, typename T2, typename Tol> constexpr bool is_greater_impl(T1 lval, T2 rval, tolerance_val<Tol> tol) {using std::abs; return not is_near_impl(lval,rval,tol) and lval > rval;}
```